mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/event/SystemTopicBasedSystemEventService.java

```diff
@@ -0,0 +1,277 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.mqtt.support.event;
+
+import com.google.common.annotations.Beta;
+import io.streamnative.pulsar.handlers.mqtt.Connection;
+import io.streamnative.pulsar.handlers.mqtt.utils.NettyUtils;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.common.util.JsonUtil;
+import org.apache.pulsar.broker.PulsarServerException;
+import org.apache.pulsar.broker.PulsarService;
+import org.apache.pulsar.broker.systopic.SystemTopicClient;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.impl.Backoff;
+import org.apache.pulsar.client.util.RetryUtil;
+import org.apache.pulsar.common.naming.NamespaceName;
+import org.apache.pulsar.common.naming.TopicName;
+
+/**
+ * System topic based event service.
+ */
+@Slf4j
+@Beta
+public class SystemTopicBasedSystemEventService implements SystemEventService {
+
+    private final PulsarService pulsarService;
+    private final TopicName topicName;
+    private final SystemTopicClient<MqttEvent> systemTopicClient;
+    private final Map<String, SourceEvent> eventCache;
+    private final ConcurrentMap<String, String> clusterClientIds;
+    private final List<EventListener> listeners;
+
+    private volatile SystemTopicClient.Reader<MqttEvent> reader;
+    private final AtomicBoolean initReader = new AtomicBoolean(false);
+    private final AtomicInteger maxRetry = new AtomicInteger(0);
+
+    public SystemTopicBasedSystemEventService(PulsarService pulsarService) {
+        this.pulsarService = pulsarService;
+        this.topicName = TopicName.get("pulsar/system/__mqtt_event");
```

Comment:
Maybe we can extract ``pulsar/system/__mqtt_event`` to a constant. This makes it more clear that it's a system topic.